### PR TITLE
ARROW-8943: [C++][Python][Dataset] Add partitioning support to ParquetDatasetFactory

### PR DESCRIPTION
**Pull request by fsaintjacques**

### CODE REVIEW

File: python/pyarrow/dataset.py [hunk 1]
```diff
@@ -486,8 +488,13 @@ def parquet_dataset(metadata_path, schema=None, filesystem=None, format=None):
         filesystem, _ = _ensure_filesystem(filesystem)
 
     metadata_path = _normalize_path(filesystem, _stringify_path(metadata_path))
+    options = ParquetFactoryOptions(
+        partition_base_dir=partition_base_dir,
+        partitioning=partitioning
```
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:493]: Can you add here a `partitioning = _ensure_partitioning(partitioning)` before passing it to ParquetFactoryOptions? 

(that will eg make sure that `parquet_dataset(..., partitioning="hive")` will also work, similarly as for `dataset(..)`)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good! Going to test it a bit further with dask.

File: cpp/src/arrow/dataset/file_parquet.h [hunk 2]
```diff
@@ -215,6 +215,34 @@ class ARROW_DS_EXPORT ParquetFileFragment : public FileFragment {
...
+  // Example:
+  // partition_base_dir = "/dataset";
+  //
+  // - "/dataset/US/sales.csv" -> "US/sales.csv" will be given to the partitioning
+  //
+  // - "/home/john/late_sales.csv" -> Will be ignored for partition discovery.
+  //
+  // This is useful for partitioning which parses directory when ordering
+  // is important, e.g. DirectoryPartitioning.
+  std::string partition_base_dir;
```
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/file_parquet.h:243]: That's a good point. I'll follow what FileSystemFactory does with the selector base's path.

File: python/pyarrow/tests/test_dataset.py [hunk 3]
```diff
@@ -1522,19 +1522,20 @@ def _create_parquet_dataset_partitioned(root_path):
...
     root_path = tempdir / "test_parquet_dataset_factory_partitioned"
     metadata_path, table = _create_parquet_dataset_partitioned(root_path)
 
-    dataset = ds.parquet_dataset(metadata_path)
-    # TODO partition column not yet included
-    # assert dataset.schema.equals(table.schema)
+    partitioning = ds.partitioning(flavor="hive")
+    dataset = ds.parquet_dataset(metadata_path,
+                                 partitioning=partitioning,
+                                 partition_base_dir=str(root_path))
```
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply python/pyarrow/tests/test_dataset.py:1535]: It is not needed anymore and fixed :)

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the updates, looks good to me!

**Pull request closed without merging**